[APPC-5012] Wallet app login adjust to open custom chrome tab

arthur-oliveira-aptoide opened the pull request:
**What does this PR do?**
Add support to Custom Chrome Tab login. 
Allowing the user to use the already saved google login's account info of his favorite browser.

Both login types (WebView and CCT) have been refactor to a common module `login` to group common logic.
To initiate the login process from inside the application the developer must call `processLoginRequest` function to verify if the target device is capable of use cct or not, and then the appropriate login version is called for the user.

**Database changed?**

No

**Where should the reviewer start?**

- [ ] SettingsFragment.kt
- [ ] HomeManageWalletBottomSheetFragment.kt
- [ ] ProcessLogin.kt
- [ ] HasCustomChromeTabAvailable.kt
- [ ] CustomTabLoginActivity.kt
**Web Checkout**
- [ ] OneStepPaymentReceiver.kt
- [ ] Erc681Receiver.kt
- [ ] CreateWebViewPaymentOspUseCase.kt
- [ ] CreateWebViewPaymentSdkUseCase.kt

**How should this be manually tested?**

  App Version must be >= 378.
  Native login inside the application.
  Login during a payment. (using SDK and OSP)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-5012](https://aptoide.atlassian.net/browse/APPC-5012)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] Database migration (if applicable)?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass


[APPC-5012]: https://aptoide.atlassian.net/browse/APPC-5012?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

arthur-oliveira-aptoide committed:
CustomTabLoginActivity implementation.

- new Activity implementation.
- SettingsFragment calls CustomTabsIntent instead of WebViewLoginActivity

This implementation aims to replace the WebView object for the CustomTabs.

TODO:
- change the manifest to setup the correct pathPreference and scheme.
- check if the `onResume` and `onNewIntent` override are needed.
- test new login flow.

arthur-oliveira-aptoide committed:
HomeFragment update.

- OnSignIn side effect CustomTabsIntent is called.

arthur-oliveira-aptoide committed:
Minor changes on CustomTabsLogin

- State observer update.
- ViewModel update.
- Dependency injection update on CustomTabLoginVM

arthur-oliveira-aptoide committed:
Code cleanup

- duplicated code has been refactor to auxiliary function.
- documentation has been added to better logical comprehension.

arthur-oliveira-aptoide committed:
ProcessLogin implementation.

- Verify if the user is capable of running cct and if it's not, calls the Login webview.

arthur-oliveira-aptoide committed:
Code refactor.

webview_login --> login.webview_login
custom_tab_login --> login.custom_tab_login

arthur-oliveira-aptoide committed:
CustomTabLogin style update.

arthur-oliveira-aptoide committed:
code cleanup

arthur-oliveira-aptoide committed:
CustomTabLoginActivity update

is_payment_in_process has been added to deeplink to when `WebViewPaymentActivity` uses the deeplink.

arthur-oliveira-aptoide committed:
Merge branch 'dev' of github.com:Catappult/appcoins-wallet-android into task/APPC-5012-Wallet-App-Log_In-Adjust-to-open-Custom-Chrome-Tab

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/com/asfoundation/wallet/ui/webview_payment/WebViewPaymentViewModel.kt

arthur-oliveira-aptoide committed:
CustomTabLoginActivity update.

PaymentOverlayHandle is being called to correctly navigate to `WebViewPaymentActivity`

arthur-oliveira-aptoide committed:
HomeManageWalletBottomSheetFragment update.

OpenLogin side effect call processLoginRequest instead of launchCustomChromeTabIntent.

arthur-oliveira-aptoide committed:
hasCustomChromeTabAvailable function implementation.

checks if the user is eligible to use custom chrome tabs.

arthur-oliveira-aptoide committed:
CreateWebViewPayment Ops and Sdk UseCase update.

both transaction types verify the web should handle a possible login in that transaction with cct or not.

arthur-oliveira-aptoide committed:
processLogin minor changes.

function refactor.
url build:
'is_cct' and 'version' parameters are being added to url.

arthur-oliveira-aptoide committed:
code minor changes

arthur-oliveira-aptoide committed:
PaymentOverlayHandler Update

The “bringToFrontAndDeliver” function has been updated to include a new parameter, “clearTop.” This parameter allows the function to clear the top element of the activity stack before bringing the "WebViewPaymentActivity" to the front. This update addresses the use case of destroying the CustomTabActivity after the login flow, specifically when using Custom Chrome Tab.

arthur-oliveira-aptoide committed:
Version Code fix.

arthur-oliveira-aptoide committed:
CustomTabLoginActivity update.

bringToFrontAndDeliver is using clearTop .

arthur-oliveira-aptoide committed:
ReLaunch WebView Payment implementation.

The ReLaunch logic has been updated to give the responsibility of starting the WebViewPaymentActivity to OneStepPaymentReceiver to avoid inconsistent state after successful login flow.

arthur-oliveira-aptoide committed:
[APPC-5012] WebViewResult Implementation

An enumeration responsible for defining the possible outcomes of a WebView.

arthur-oliveira-aptoide committed:
[APPC-5012] CreateWebViewPaymentSdkUseCase update

url build has been fixed.

arthur-oliveira-aptoide committed:
[APPC-5012] WebViewPaymentActivity update.

restartWebViewPayment's logic update:
	Each receiver is responsible for relaunching the WebViewPaymentActivity.

arthur-oliveira-aptoide committed:
[APPC-5012] Minor Modifications

Code errors have been fixed.

rodolfobcardoso commented on app/src/main/java/com/asfoundation/wallet/ui/webview_payment/WebViewPaymentViewModel.kt:
```diff
@@ -273,7 +273,12 @@ class WebViewPaymentViewModel @Inject constructor(
+    transaction: TransactionBuilder,
+    context: Context
+  ) {
     CompositeDisposable().add(
```
It's not from your PR, but this CompositeDisposable().add() is not correct. It should be added to a common object and then disposed of by calling the disposables.clear() in the onDestroyView.
Can you please fix?

rodolfobcardoso commented on app/src/main/java/com/asfoundation/wallet/ui/settings/entry/SettingsFragment.kt:
```diff
@@ -13,6 +13,7 @@ import android.widget.FrameLayout
 import android.widget.Toast
 import androidx.activity.result.ActivityResultLauncher
 import androidx.activity.result.contract.ActivityResultContracts
+import androidx.browser.customtabs.CustomTabsIntent
```
Can be removed

rodolfobcardoso commented on app/src/main/java/com/asfoundation/wallet/ui/webview_payment/WebViewPaymentActivity.kt:
```diff
@@ -142,7 +144,7 @@ class WebViewPaymentActivity : AppCompatActivity() {
     super.onNewIntent(intent)
     val data = intent?.data?.toString().orEmpty()
-
+    Log.d(TAG, "onNewIntent: $data")
```
For safety it's better to remove this log

rodolfobcardoso commented on app/src/main/java/com/asfoundation/wallet/ui/webview_payment/usecases/CreateWebViewPaymentOspUseCase.kt:
```diff
@@ -35,7 +37,8 @@ class CreateWebViewPaymentOspUseCase @Inject constructor(
     transaction: TransactionBuilder,
-    appVersion: String?
+    appVersion: String?,
+    context: Context
```
Context should not be used in the domain layer (use cases), otherwise the use case won't be testable.
It's better to pass the variable you need (for example hasCustomTab: Boolean) instead of the context.

rodolfobcardoso commented on app/src/main/java/com/asfoundation/wallet/ui/webview_payment/usecases/CreateWebViewPaymentSdkUseCase.kt:
```diff
@@ -35,7 +37,8 @@ class CreateWebViewPaymentSdkUseCase @Inject constructor(
     transaction: TransactionBuilder,
-    appVersion: String?
+    appVersion: String?,
+    context: Context
```
Same here

rodolfobcardoso reviewed (changes requested):
Left some notes in the code.

Also, please check again the login flow from inside the wallet, seems like the deeplink is not being called.

arthur-oliveira-aptoide committed:
[APPC-5012] Refactor CreateWebViewPaymentOspUseCase and CreateWebViewPaymentSdkUseCase

- Renamed “isCct” to “hasCustomTab” for compression enhancement.
- Removed “context” from the function parameter to improve modularization and testability.

arthur-oliveira-aptoide committed:
[APPC-5012] Code cleanup

arthur-oliveira-aptoide committed:
[APPC-5012] WebViewPaymentViewModel update

- `fetchUserKey` function refactor:
	CompositeDisposable -> compositeDisposable.
	Shared object used inside the viewModel, this ensures that when `onDestroyView` is called all disposables are cleared.

arthur-oliveira-aptoide committed:
[APPC-5012]  code cleanup

unnecessary logger has been removed.

arthur-oliveira-aptoide committed:
[APPC-5012] NativeLoginActivity implementation.

This activity is used to be an anchor to close the `CustomTabsIntent` Activity before processing the login request.
This is necessary to avoid being stuck in the user's google account while processing the login information.
With this Activity the user see `MainActivity` instead, improving the user experience.

arthur-oliveira-aptoide committed:
[5012] Fix Version Code

arthur-oliveira-aptoide committed:
[APPC-5012] CustomTabLoginActivity update

- repeatOnLifecycle has been added to `lifecycleScope` to properly handle lifecycle events.

arthur-oliveira-aptoide committed:
[APPC-5012] code refactor

arthur-oliveira-aptoide committed:
[APPC-5012] email has been added to WebViewPaymentInterface
